subsys/bluetooth/audio/has_client.c

```diff
@@ -82,11 +88,195 @@ static enum bt_has_capabilities get_capabilities(const struct has_inst *inst)
 	return caps;
 }
 
+union preset_record {
+	struct bt_has_preset_record record;
+	struct {
+		enum bt_has_properties properties;
+		char name[BT_HAS_PRESET_NAME_MAX + 1]; /* + 1 byte for null-terminator */
+	};
+};
+
+static void handle_read_preset_rsp(struct bt_conn *conn, struct net_buf_simple *buf)
+{
+	struct has_inst *inst = inst_by_conn(conn);
+	const struct bt_has_cp_read_preset_rsp *rsp;
+	union preset_record record;
+	size_t name_len;
+
+	BT_DBG("conn %p buf %p", (void *)conn, buf);
+
+	if (buf->len < sizeof(*rsp)) {
+		BT_ERR("malformed PDU");
+		return;
+	}
+
+	rsp = net_buf_simple_pull_mem(buf, sizeof(*rsp));
+
+	record.properties = rsp->properties;
+	name_len = MIN(buf->len + 1, ARRAY_SIZE(record.name));
+	utf8_lcpy(record.name, rsp->name, name_len);
+
+	if (rsp->is_last > 0x01) {
+		BT_WARN("unexpected is_last 0x%02x", rsp->is_last);
+	}
+
+	client_cb->preset_event(&inst->has, 0, BT_HAS_PRESET_RECORD_READ_RSP, rsp->index,
+				&record.record, !!rsp->is_last);
+}
+
+static void handle_generic_update(struct bt_conn *conn, struct net_buf_simple *buf, bool is_last)
+{
+	struct has_inst *inst;
+	const struct bt_has_cp_generic_update *generic_update;
+	struct bt_has_preset_delete delete;
+	union preset_record record;
+	size_t name_len;
+
+	if (buf->len < sizeof(*generic_update)) {
+		BT_ERR("malformed PDU");
+		return;
+	}
+
+	inst = inst_by_conn(conn);
+
+	generic_update = net_buf_simple_pull_mem(buf, sizeof(*generic_update));
+
+	/* Generic Update can be used as well to notify preset deletion.
+	 * Notify the user that there are no presets registered in range (PrevIndex, Index)
+	 */
+	delete.end_index = generic_update->index - 1;
+	client_cb->preset_event(&inst->has, 0, BT_HAS_PRESET_DELETE,
+				generic_update->prev_index + 1, &delete, false);
```

Comment:
Given that the generic update does not necessarily mean that a preset was removed (in fact the spec states that it `The Generic Update ChangeId shall be used to inform clients when a new preset record is added to the list of preset records`) we shouldn't call the `preset_event` with `BT_HAS_PRESET_DELETE` unconditionally. 

Can we check if there is a deletion by checking the cache?

Reply:
The question is where shall be the cache, what should be it's size and how the cache should look like (should we cache every preset or the application should decide which presets it wants to cache).

The problem is this is implementation specific :(

If you have an idea how it should look like I would be happy to hear because I do not know where to put it and how to handle preset changes. I favor an option that the application takes care of that

Reply:
Since the HAS client have been implemented with the idea that the data lives in the application, then perhaps we should simply just forward the notifications/indications directly to the application (with verification of the values of course), instead of having the `preset_event`? 

If the data and logic exists in the application layer, then I guess the stack shouldn't try to be smart and simply act as a layer that does minimum verification (and of course the translations from `bt_conn` to `bt_has`).

Reply:
OK, done